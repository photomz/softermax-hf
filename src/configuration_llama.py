"""
Subclasses the original llama config to add softermax denominator n_bias
"""

from transformers.models.llama import LlamaConfig


class SofterLlamaConfig(LlamaConfig):
    def __init__(self, n_bias: float = 1.0, **kwargs):
        super().__init__(**kwargs)
        self.n_bias = n_bias
