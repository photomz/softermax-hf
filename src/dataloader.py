import itertools
import os
from dataclasses import dataclass, field
from typing import Dict, Literal, Optional
from functools import cached_property

import torch
from datasets import IterableDatasetDict, concatenate_datasets, load_dataset
from torch.utils.data import DataLoader
from transformers import (
    PreTrainedTokenizerBase,
    DataCollatorForLanguageModeling,
    DataCollatorWithPadding,
    DefaultDataCollator,
)


@dataclass
class BooksCorpusAndWiki:
    tokenizer: PreTrainedTokenizerBase

    # stream shuffled dataset by batches of buffer_size
    buffer_size: int = 30000
    max_seq_length: int = 128
    batch_size: Dict[str, int] = field(default_factory=lambda: {"train": 32, "validation": 32})

    # % tokens to mask for Encoders. Set to 0% to disable for Decoders.
    mlm_probability: float = 0.15
    # random seed for shuffling and masking.
    seed: int = 42
    # HuggingFace iterable dataset dict.
    datasets: Optional[IterableDatasetDict] = None

    # no. cpu threads. Defaults to # cores - 2.
    # these are used to determine how many threads will process the streaming
    # dataset concurrently with model training
    num_workers: int = os.cpu_count() - 2

    def setup(self, suffix=""):
        """Streams, shuffles, tokenizes, and groups the datasets.

        Args:
            suffix (str, optional): % of datasplit. Defaults to 100%. Format: [:n%]
        """

        # stream the dataset as it is too large to download
        bookcorpus = load_dataset("bookcorpus", split=f"train{suffix}", streaming=True)
        wiki_train = load_dataset("wiki40b", "en", split=f"train{suffix}", streaming=True)
        wiki_val = load_dataset("wiki40b", "en", split=f"validation{suffix}", streaming=True)

        # keep text column to concat with bookcorpus
        wiki_train = wiki_train.remove_columns([col for col in wiki_train.column_names if col != "text"])
        wiki_val = wiki_val.remove_columns([col for col in wiki_val.column_names if col != "text"])

        assert bookcorpus.features.type == wiki_train.features.type

        # dataset transform steps
        self.datasets = (
            IterableDatasetDict({"train": concatenate_datasets([bookcorpus, wiki_train]), "validation": wiki_val})
            .shuffle(seed=self.seed, buffer_size=self.buffer_size)
            .map(self.encode, batched=True, remove_columns=wiki_val.column_names)
            .map(self.group, batched=True)
            .with_format("torch")
        )

        assert isinstance(self.datasets["train"], torch.utils.data.IterableDataset)
        assert isinstance(self.datasets["validation"], torch.utils.data.IterableDataset)

    @cached_property
    def training_args(self):
        """
        Dataset-specific **kwargs to init HuggingFace TrainingArguments
        Ref: https://huggingface.co/docs/transformers/v4.36.1/en/main_classes/trainer#transformers.TrainingArguments
        """
        return {
            "dataloader_num_workers": self.num_workers,
            "data_seed": self.seed,
        }

    @cached_property
    def trainer_params(self):
        """
        Dataset-specific **kwargs to init HF Trainer
        Ref: https://huggingface.co/docs/transformers/main_classes/trainer#transformers.Trainer
        """
        # we initialise different data collators based on causal or masked language modeling
        if self.mlm_probability:
            # in MLM we usually only have [CLS] and [SEP] tokens
            collate_fn = DataCollatorForLanguageModeling(self.tokenizer, mlm=True, mlm_probability=self.mlm_probability)
        else:
            # DataCollatorForLanguageModeling(mlm=False) requires a pad_token (which doesn't exist) to be set, but
            # our method of CLM pretraining we chunk the input (see group function) for pretraining with large amounts
            # of data, hence no [PAD] token. This means the outputs of the dataloader are all `max_seq_len` long.
            # So here we explicitly use the default data collator which doesn't include any padding, outputting samples as is.
            # The caveat here is that we have to manually create the 'labels' column, which we do at the end of .group()
            collate_fn = DefaultDataCollator(return_tensors="pt")

        return {
            "train_dataset": self.datasets["train"],
            "eval_dataset": self.datasets["validation"],
            "data_collator": collate_fn,
        }

    @cached_property
    def calibration_split(self):
        """
        Dataset-specific calibration samples for GPTQConfig, returned as list of strings (since that's the format they require)
        The calibration split is equal to the validation split. Important to be a cached_property since it's rather
        expensive to turn the whole eval dataset into a list every time the getter is called.
        """
        return list(self.datasets["validation"].take(self.quant_dataset_size))

    @property
    def quant_dataset_size(self):
        # quantization calibration dataset size
        # original GPTQ paper used 128 random 2048 token segments from the C4 dataset
        # we match the same token count calculating using max_seq_len parameter
        return int(128 * 2048 / self.max_seq_length)

    def dataloader(self, split: Literal["train", "validation"] = "train"):
        torch.manual_seed(self.seed)
        # we initialise different data collators based on causal or masked language modeling
        if self.mlm_probability:
            # in MLM we usually only have [CLS] and [SEP] tokens
            collate_fn = DataCollatorForLanguageModeling(self.tokenizer, mlm=True, mlm_probability=self.mlm_probability)
        else:
            collate_fn = DefaultDataCollator(return_tensors="pt")
        return DataLoader(
            self.datasets[split],
            batch_size=self.batch_size[split],
            num_workers=self.num_workers,
            collate_fn=collate_fn,
        )

    def group(self, example_batch):
        """
        Concatenate the dataset into chunks of max_seq_length. Don't waste tokens exceeding seqlen.
        Ref: https://huggingface.co/learn/nlp-course/chapter7/3?fw=pt#preprocessing-the-data
        """
        # Flatten batch
        concatenated_examples = {k: list(itertools.chain(*example_batch[k])) for k in example_batch.keys()}
        total_length = len(concatenated_examples[list(example_batch.keys())[0]])
        # We drop the small remainder, we could add padding if the model supported it instead of
        # this drop, you can customize this part to your needs.
        if total_length >= self.max_seq_length:
            total_length = (total_length // self.max_seq_length) * self.max_seq_length
        # Split by chunks of max_len.
        result = {
            k: [t[i : i + self.max_seq_length] for i in range(0, total_length, self.max_seq_length)]
            for k, t in concatenated_examples.items()
        }

        # create new labels column
        # remember the model will automatically right-shift them for us when calculating loss
        result["labels"] = result["input_ids"].copy()
        return result

    def encode(self, example_batch):
        """
        Tokenizes all the text into tokens first, the chunking will be done after by .group() function
        so we don't unnecessarily drop tokens by using huggingface's built-in chunking method.
        Ref: https://huggingface.co/learn/nlp-course/chapter7/6?fw=pt#preparing-the-dataset
        """
        # tokenize the text
        features = self.tokenizer(
            example_batch["text"],
            return_special_tokens_mask=True,
        )
        return features
