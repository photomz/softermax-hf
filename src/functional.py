"""
softermax implementation
"""

from typing import Optional

import torch
import torch.nn.functional as F
from torch.types import _dtype


# Ref: https://github.com/softmax1/EsperBERTo/blob/946f0a9fa0f6e3b2bf755388d8fa20c31f8e2bf5/src/functional.py#L49
def softermax(
    input: torch.Tensor, n_bias: float = 0.0, dim: Optional[int] = -1, dtype: Optional[_dtype] = None
) -> torch.Tensor:
    """
    $\text(softmax)_n(x_i) = exp(x_i) / (n + \sum_j exp(x_j))$

    Note: softmax_n, with fixed input, is _not_ shift-symmetric when n != 0, and we must account for this.
    Normally when computing a softmax, the maxes are subtracted from the inputs for numeric stability.
    """
    # compute the maxes along the last dimension
    input_maxes = input.max(dim=dim, keepdim=True).values.detach()
    # shift the input to prevent overflow (and underflow in the denominator)
    # compute the numerator and softmax_0 denominator using the shifted input
    numerator = torch.exp(input - input_maxes)
    denominator = numerator.sum(dim=dim, keepdim=True)
    # we need to shift the zeros in the same way we shifted the inputs
    shifted_bias = n_bias * torch.exp(-input_maxes)
    # and then add this contribution to the denominator
    scores = numerator / (denominator + shifted_bias)
    return scores if dtype is None else scores.type(dtype=dtype)


# Ref: https://github.com/Qualcomm-AI-research/outlier-free-transformers/blob/1e59744972da808a90d9996027bfc274689594b8/transformers_language/models/softmax.py#L8
def clipped_softmax(input: torch.Tensor, dim: int = -1, gamma: float = 0.0, eta: float = 1.0, **kwargs) -> torch.Tensor:
    """clip((eta - gamma) * softmax(x) + gamma) in [0,1], eta >= 1, gamma <= 0.

    Exact zeros can fix outliers, so fix eta = 1 (no upper clipping) and vary gamma (lower clipping).
    Normal softmax is clipped_softmax(eta=1, gamma=0).
    """
    # F.softmax doesn't support fp16 on CPU
    score = F.softmax(input, dim=dim, **kwargs)
    print(score)
    # Stretch and translate by linear factor
    stretched_score = score * (eta - gamma) + gamma
    # Clip each score to [0,1], so sum != 1 anymore.
    # In fact, every score in attention sum can be 0 or 1.
    clipped_score = torch.clip(stretched_score, 0, 1)
    return clipped_score
