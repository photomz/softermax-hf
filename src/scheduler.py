"""
Implements WarmupDecayedCosineAnnealingWarmRestarts,
Run the file to see scheduler.png to see the lr graph. 
Basically linear warmup then cosine annealing with warm restarts.
"""

import torch
import math
import functools

from torch.optim.lr_scheduler import CosineAnnealingWarmRestarts, SequentialLR


class CosineAnnealingWarmRestartsDecay(CosineAnnealingWarmRestarts):
    """
    Subclass of regular cosine annealing with warm restarts to add decaying peak learning rates.
    Ref: https://stackoverflow.com/a/73747249
    """

    def __init__(self, optimizer, T_0, T_mult=1, eta_min=0, last_epoch=-1, verbose=False, decay=1):
        super().__init__(optimizer, T_0, T_mult=T_mult, eta_min=eta_min, last_epoch=last_epoch, verbose=verbose)
        self.decay = decay
        self.initial_lrs = self.base_lrs

    def step(self, epoch=None):
        if epoch == None:
            if self.T_cur + 1 == self.T_i:
                if self.verbose:
                    print("multiplying base_lrs by {:.4f}".format(self.decay))
                self.base_lrs = [base_lr * self.decay for base_lr in self.base_lrs]
        else:
            if epoch < 0:
                raise ValueError("Expected non-negative epoch, but got {}".format(epoch))
            if epoch >= self.T_0:
                if self.T_mult == 1:
                    n = int(epoch / self.T_0)
                else:
                    n = int(math.log((epoch / self.T_0 * (self.T_mult - 1) + 1), self.T_mult))
            else:
                n = 0

            self.base_lrs = [initial_lrs * (self.decay**n) for initial_lrs in self.initial_lrs]

        super().step(epoch)


def _constant_warmup(iteration, warmup_iterations):
    """
    Linear warmup from 0 --> 1.0, then constant
    """
    multiplier = 1.0
    if iteration <= warmup_iterations:
        multiplier = iteration / warmup_iterations
    return multiplier


def LinearWarmup(optimizer, T_warmup):
    _decay_func = functools.partial(_constant_warmup, warmup_iterations=T_warmup)
    scheduler = torch.optim.lr_scheduler.LambdaLR(optimizer, _decay_func)
    return scheduler


def WarmupDecayedCosineAnnealingWarmRestarts(optimizer, warmup_iters, T_0=500, decay=0.8):
    """
    Adds a linear warmup in front of the decayed cosine annealing with warm restarts

    optimizer: self-explanatory
    warmup_iters: number of iterations for the linear warmup
    T_0: Number of iterations for the first restart
    decay: decay factor for each warm restart, restarted_peak_lr = prev_peak_lr * decay
    """
    warmup_scheduler = LinearWarmup(optimizer, T_warmup=warmup_iters)
    cos_scheduler = CosineAnnealingWarmRestartsDecay(optimizer, T_0=T_0, decay=decay)
    scheduler = SequentialLR(optimizer, schedulers=[warmup_scheduler, cos_scheduler], milestones=[warmup_iters])
    return scheduler


if __name__ == "__main__":

    import matplotlib

    matplotlib.use("Agg")
    import matplotlib.pyplot as plt

    # Dummy parameters
    parameters = torch.tensor([0.0, 0.0, 0.0], requires_grad=True)
    total_iters = 2000
    warmup_iters = 100

    # Test
    optimizer = torch.optim.Adam([parameters], lr=0.2)
    scheduler = WarmupDecayedCosineAnnealingWarmRestarts(optimizer, warmup_iters, T_0=300, decay=0.8)

    actual_lr = []
    for _iter in range(total_iters):
        optimizer.step()
        scheduler.step()
        actual_lr.append(optimizer.param_groups[0]["lr"])
    plt.plot(list(range(total_iters)), actual_lr, "-", label="CosineAnnealingWarmRestartsDecay with Warmup")

    plt.xlabel("iterations")
    plt.ylabel("learning rate")
    plt.legend()
    plt.savefig("scheduler.png")
