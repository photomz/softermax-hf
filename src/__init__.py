"""
__init__ file, pretty straightforward
"""

from .configuration_llama import SofterLlamaConfig
from .modeling_llama import (
    SofterLlamaPreTrainedModel,
    SofterLlamaModel,
    SofterLlamaForCausalLM,
)
