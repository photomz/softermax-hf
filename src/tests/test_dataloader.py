import pytest
from transformers import AutoModel, RobertaTokenizer, Trainer, TrainingArguments

from src.dataloader import BooksCorpusAndWiki


# These tests require Internet to download the dataset, or a precached dataset.
@pytest.fixture(scope="session")
def bookscorpusandwiki():
    tokenizer = RobertaTokenizer.from_pretrained("roberta-base")
    loader = BooksCorpusAndWiki(tokenizer)
    loader.setup()
    return loader


@pytest.fixture(scope="session")
def model():
    return AutoModel.from_pretrained("prajjwal1/bert-tiny")


def test_trainset_bookcorpusnwiki(bookscorpusandwiki, snapshot):
    """first batch of trainset (BooksCorpusAndWiki) should match snapshot"""
    trainset = bookscorpusandwiki.datasets["train"]
    next_batch = next(iter(trainset))
    assert all(key in next_batch.keys() for key in ["input_ids", "attention_mask", "special_tokens_mask"])
    assert next_batch["input_ids"].shape == snapshot
    assert next_batch == snapshot


def test_trainloader_bookcorpusnwiki(bookscorpusandwiki, snapshot):
    """first batch of trainloader (BooksCorpusAndWiki) should match snapshot"""
    trainloader = bookscorpusandwiki.dataloader("train")
    next_batch = next(iter(trainloader))
    assert all(key in next_batch.keys() for key in ["input_ids", "attention_mask", "labels"])
    assert next_batch["input_ids"].shape == snapshot
    assert next_batch == snapshot


def test_trainingargs_bookcorpusnwiki(bookscorpusandwiki, model, snapshot):
    """training arguments for BooksCorpusAndWiki should match snapshot"""
    required_args = {"output_dir": "/tmp", "max_steps": int(1e5), "logging_dir": "/tmp/runs"}
    args = TrainingArguments(**bookscorpusandwiki.training_args, **required_args)
    assert args == snapshot

    # Don't snapshot-test Trainer. It has hardware-specific configs, so it's not portable.
    assert Trainer(model, args, **bookscorpusandwiki.trainer_params)
