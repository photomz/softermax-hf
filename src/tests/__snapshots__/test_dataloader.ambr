# serializer version: 1
# name: test_trainingargs_bookcorpusnwiki
  TrainingArguments(
  _n_gpu=1,
  adafactor=False,
  adam_beta1=0.9,
  adam_beta2=0.999,
  adam_epsilon=1e-08,
  auto_find_batch_size=False,
  bf16=False,
  bf16_full_eval=False,
  data_seed=42,
  dataloader_drop_last=False,
  dataloader_num_workers=6,
  dataloader_persistent_workers=False,
  dataloader_pin_memory=True,
  ddp_backend=None,
  ddp_broadcast_buffers=None,
  ddp_bucket_cap_mb=None,
  ddp_find_unused_parameters=None,
  ddp_timeout=1800,
  debug=[],
  deepspeed=None,
  disable_tqdm=False,
  dispatch_batches=None,
  do_eval=False,
  do_predict=False,
  do_train=False,
  eval_accumulation_steps=None,
  eval_delay=0,
  eval_steps=None,
  evaluation_strategy=IntervalStrategy.NO,
  fp16=False,
  fp16_backend=auto,
  fp16_full_eval=False,
  fp16_opt_level=O1,
  fsdp=[],
  fsdp_config={'min_num_params': 0, 'xla': False, 'xla_fsdp_grad_ckpt': False},
  fsdp_min_num_params=0,
  fsdp_transformer_layer_cls_to_wrap=None,
  full_determinism=False,
  gradient_accumulation_steps=1,
  gradient_checkpointing=False,
  gradient_checkpointing_kwargs=None,
  greater_is_better=None,
  group_by_length=False,
  half_precision_backend=auto,
  hub_always_push=False,
  hub_model_id=None,
  hub_private_repo=False,
  hub_strategy=HubStrategy.EVERY_SAVE,
  hub_token=<HUB_TOKEN>,
  ignore_data_skip=False,
  include_inputs_for_metrics=False,
  include_num_input_tokens_seen=False,
  include_tokens_per_second=False,
  jit_mode_eval=False,
  label_names=None,
  label_smoothing_factor=0.0,
  learning_rate=5e-05,
  length_column_name=length,
  load_best_model_at_end=False,
  local_rank=0,
  log_level=passive,
  log_level_replica=warning,
  log_on_each_node=True,
  logging_dir=/tmp/runs,
  logging_first_step=False,
  logging_nan_inf_filter=True,
  logging_steps=500,
  logging_strategy=IntervalStrategy.STEPS,
  lr_scheduler_kwargs={},
  lr_scheduler_type=SchedulerType.LINEAR,
  max_grad_norm=1.0,
  max_steps=100000,
  metric_for_best_model=None,
  mp_parameters=,
  neftune_noise_alpha=None,
  no_cuda=False,
  num_train_epochs=3.0,
  optim=OptimizerNames.ADAMW_TORCH,
  optim_args=None,
  output_dir=/tmp,
  overwrite_output_dir=False,
  past_index=-1,
  per_device_eval_batch_size=32,
  per_device_train_batch_size=32,
  prediction_loss_only=False,
  push_to_hub=False,
  push_to_hub_model_id=None,
  push_to_hub_organization=None,
  push_to_hub_token=<PUSH_TO_HUB_TOKEN>,
  ray_scope=last,
  remove_unused_columns=True,
  report_to=[],
  resume_from_checkpoint=None,
  run_name=/tmp,
  save_on_each_node=False,
  save_only_model=False,
  save_safetensors=True,
  save_steps=500,
  save_strategy=IntervalStrategy.STEPS,
  save_total_limit=None,
  seed=42,
  skip_memory_metrics=True,
  split_batches=False,
  tf32=None,
  torch_compile=False,
  torch_compile_backend=None,
  torch_compile_mode=None,
  torchdynamo=None,
  tpu_metrics_debug=False,
  tpu_num_cores=None,
  use_cpu=False,
  use_ipex=False,
  use_legacy_prediction_loop=False,
  use_mps_device=False,
  warmup_ratio=0.0,
  warmup_steps=0,
  weight_decay=0.0,
  )
# ---
# name: test_trainloader_bookcorpusnwiki
  Size(
    32,
    128,
  )
# ---
# name: test_trainloader_bookcorpusnwiki.1
  {'input_ids': tensor([[    0,  8877,   128,  ...,     1,     1,     1],
          [    0,    90, 12541,  ...,     1,     1,     1],
          [    0, 50264,    42,  ...,     1,     1,     1],
          ...,
          [    0,   627,  4382,  ...,     1,     1,     1],
          [    0, 49519,   939,  ...,     1,     1,     1],
          [    0, 49519, 14223,  ...,     1,     1,     1]]), 'attention_mask': tensor([[1, 1, 1,  ..., 0, 0, 0],
          [1, 1, 1,  ..., 0, 0, 0],
          [1, 1, 1,  ..., 0, 0, 0],
          ...,
          [1, 1, 1,  ..., 0, 0, 0],
          [1, 1, 1,  ..., 0, 0, 0],
          [1, 1, 1,  ..., 0, 0, 0]]), 'labels': tensor([[ -100,  -100,  -100,  ...,  -100,  -100,  -100],
          [ -100,  -100,  -100,  ...,  -100,  -100,  -100],
          [ -100, 49519,  -100,  ...,  -100,  -100,  -100],
          ...,
          [ -100,  -100,  -100,  ...,  -100,  -100,  -100],
          [ -100,  -100,  -100,  ...,  -100,  -100,  -100],
          [ -100,  -100,  -100,  ...,  -100,  -100,  -100]])}
# ---
# name: test_trainset_bookcorpusnwiki
  Size(
    128,
  )
# ---
# name: test_trainset_bookcorpusnwiki.1
  dict({
    'attention_mask': tensor([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0,
          0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
          0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
          0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
          0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
          0, 0, 0, 0, 0, 0, 0, 0]),
    'input_ids': tensor([    0,  8877,   128,   417,   393,    56,    10,   313, 15256,   215,
           5676,  6453,    13,    69,   479,     2,     1,     1,     1,     1,
              1,     1,     1,     1,     1,     1,     1,     1,     1,     1,
              1,     1,     1,     1,     1,     1,     1,     1,     1,     1,
              1,     1,     1,     1,     1,     1,     1,     1,     1,     1,
              1,     1,     1,     1,     1,     1,     1,     1,     1,     1,
              1,     1,     1,     1,     1,     1,     1,     1,     1,     1,
              1,     1,     1,     1,     1,     1,     1,     1,     1,     1,
              1,     1,     1,     1,     1,     1,     1,     1,     1,     1,
              1,     1,     1,     1,     1,     1,     1,     1,     1,     1,
              1,     1,     1,     1,     1,     1,     1,     1,     1,     1,
              1,     1,     1,     1,     1,     1,     1,     1,     1,     1,
              1,     1,     1,     1,     1,     1,     1,     1]),
    'special_tokens_mask': tensor([1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1,
          1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
          1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
          1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
          1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
          1, 1, 1, 1, 1, 1, 1, 1]),
  })
# ---
